tf2onnx/rewriter/gemm_rewriter.py

```diff
@@ -65,140 +69,70 @@ def rewrite_gemm(g, ops):
-            inputC_node = match.get_op("C")
-
-            # for edges:
-            a_edge_name = _find_edges_name_btw_nodes(inputA_node, matmul_node)
-            b_edge_name = _find_edges_name_btw_nodes(inputB_node, matmul_node)
-            c_edge_name = _find_edges_name_btw_nodes(inputC_node, add_node)
-
-            # alpha = match.get_op("alpha").get_tensor_value()
-            # beta = match.get_op("beta").get_tensor_value()
-
-            gemm = g.make_node("Gemm", inputs=[a_edge_name, b_edge_name, c_edge_name],
-                               shapes=[g.get_shape(add_node.output[0])], dtypes=[g.get_dtype(add_node.output[0])]) # add_node.output_shapes[0]
-
-            ops.remove(add_node)
-            ops.remove(matmul_node)
-
-    ops.append(gemm)
-    g.replace_all_inputs(ops, add_node.output[0], gemm.output[0])
-
-    return ops
-
-def _find_edges_name_btw_nodes(sender, sinker):
-    for sinker_end in sinker.input:
-        for sender_end in sender.output:
-            if sinker_end == sender_end:
-                return sinker_end
-    return None
\ No newline at end of file
+            for match in match_results:
+                add_node = match.get_op('add')
+                matmul_node = match.get_op("matmul")
+                inputA_node = match.get_op("A")
+                inputB_node = match.get_op("B")
+                inputC_node = match.get_op("C")
+                a_edge_name = inputA_node.output[0]
+                b_edge_name = inputB_node.output[0]
+                c_edge_name = inputC_node.output[0]
+                attr4makenode = {}
+
+                if patternID == 0 or patternID ==4:  # pattern 0 and 4: alpha*A*B + beta*C and beta*C + alpha*A*B
+                    mul2_node = match.get_op("mul2")
+                    mul1_node = match.get_op("mul1")
+                    alpha = match.get_op("alpha").get_tensor_value()
+                    beta = match.get_op("beta").get_tensor_value()
+                    attr4makenode = {"alpha": alpha, "beta": beta}
+
+                if patternID == 1:  # pattern1: alpha*A*B + C
+                    mul1_node = match.get_op("mul1")
+                    alpha = match.get_op("alpha").get_tensor_value()
+                    attr4makenode = {"alpha": alpha}
+
+                if patternID == 2:  # pattern2: A*B + beta*C
+                    mul2_node = match.get_op("mul2")
+                    beta = match.get_op("beta").get_tensor_value()
+                    attr4makenode = attr={"beta": beta}
+
+                # if the pattern is 3, do nothing
+                gemm = g.make_node("Gemm", inputs=[a_edge_name, b_edge_name, c_edge_name],
+                                   attr=attr4makenode,
+                                   # outputs=[add_node.output[0]],
+                                   shapes=[g.get_shape(add_node.output[0])],
+                                   dtypes=[g.get_dtype(add_node.output[0])])
+                ops.remove(add_node)
```

Comment:
Optional: no need to remove add node since you replace other nodes' input from it. these unused nodes will be eliminated by delete_unused_nodes.